### DISCUSSION TITLE
Diagnosing tests locking up

### DISCUSSION THREAD

**Discussion by bforl** [xunit/xunit#3061] [Question] [answered]
I'm seeing an odd intermittent problem, where sometimes, a number of tests will lock up indefinitely.

When I manage to get them to lock up in Visual Studio, I then open a second instance of visual studio and attach to the test host process and then pause all threads.

When I do this, I can see that a lot of the tests that have locked up are all stuck  on a line that looks like this

```
public class
{
    internal Other Other{ get; private set; } = new(); <-- The statement selects the 'new()' part as though its stuck trying to create it
}
```

where this class is just data with a static constructor (no normal constructor)

```
public class Other
{
    private static readonly int s_count = 0;

    static Other()
    {
        ... initialise simple fields
        Interlocked.Increment(ref s_count);
    }

    ... some static fields (strings, floats etc)
}
```

I can't make any sense of why it would lock up here. There is no locking in this class, its just simple data.
If I check the `s_count` field in one of the locked up threads, it's 1. 

One note is that we're running all the tests in parallel, but as far as I know, `static constructors` are perfectly safe and are locked safely by the language, so they're only called once.

Any thoughts on how I can investigate this?

Another point is that when this lock up occurs **ALL** tests in the process lock up, suspiciously in constructors. Even those that have nothing to do with the above code. If I continue the threads and then pause again, none of the tests have made any progress, even the unrelated tests. So it feels as though the whole process, or at least all tasks running tests are stuck.

Could it be that GC is stuck? I don't see any finaliser threads when I pause when attached.

C#, .NET 8, XUnit 2.9.2

**Comment by bradwilson** [comment, marked as answer]
This is what the [static constructor documentation](https://learn.microsoft.com/dotnet/csharp/programming-guide/classes-and-structs/static-constructors#remarks) has to say about parallelism:

> The runtime calls a static constructor no more than once in a single application domain. That call is made in a locked region based on the specific type of the class. No extra locking mechanisms are needed in the body of a static constructor. To avoid the risk of deadlocks, don't block the current thread in static constructors and initializers. For example, don't wait on tasks, threads, wait handles or events, don't acquire locks, and don't execute blocking parallel operations such as parallel loops, Parallel.Invoke and Parallel LINQ queries.

`Interlocked.Increment` is a locking mechanism. What is the purpose of the counter, and why must it be locked? Have you tried removing the lock?

(Personally, I don't know _why_ this would be a problem, but I'm grasping a bit at straws based on your incomplete information provided. A full repro would allow me to dig in much deeper to understand possibly what's going on.)

**Reply by bforl** [reply]
Apologies, forgot to mention that I added the interlocked increment to try and debug the issue in question. It was so I could see that the count never goes above 1.

I will try and create a repro case from scratch, but I fear it will be difficult as its intermittent, and only seems to occur when running all our other tests at the same time.

Just to clarify. Am I right in thinking it could only be tests within the same .csproj that could possibly be related to the issue? And that running other tests at the same time from another csproj couldn't interfere as they're in a different test host process?

**Reply by bradwilson** [reply]
> Am I right in thinking it could only be tests within the same .csproj that could possibly be related to the issue? And that running other tests at the same time from another csproj couldn't interfere as they're in a different test host process?

Not necessarily. It's up to VSTest, which is the ultimate test host, to decide whether two test projects running at the same time live in the same process space. Since you're using v2, which is by definition an in-process model, then they very well could be running in the same process space. If you use v3, then they will definitely be running in separate processes, because in v3 test projects are processes, not libraries.

**Reply by bradwilson** [reply]
One thing that I do know for sure: being locked waiting for the `new()` statement to run (i.e., you aren't even in the `Other` constructor) is almost assuredly a deadlock waiting for static initialization. I can't think of many other reasons why you'd be stuck there, as opposed to inside the actual constructor...assuming you even have a constructor. 😄

**Reply by bradwilson** [reply]
It's worth doubling checking `... initialise simple fields` to make sure there's nothing async happening in there (especially nothing locked behind something like `.Result` or `.GetAwaiter().GetResult()` on a task).

**Reply by bforl** [reply]
Nope, nothing like that. Just some setup of primitive types. 

If these other threads were locked up waiting for static initialisation, wouldn't I see another thread IN the static initialisation? But I see nothing like that in WinDbg.

Could it be somehow that the thread doing the static initialisation died whilst having the lock? But then I don't see how the code could throw as its not doing anything complicated.

Could it be that xunit could have killed the thread whilst in the static constructor for some reason? 

The confusing thing is why it also freezes/locks up completely unrelated tests though. 🤔

**Reply by bradwilson** [reply]
We can't kill threads in .NET 8; the API isn't present. (You can kill threads in .NET Framework, but they removed that functionality when moving to .NET Core because it's dangerous and often leaves the process is a potentially broken state).

The fact that unrelated tests are frozen would make sense if enough tests are blocking waiting for the lock.

An obvious way to verify that it's a deadlock would be to run with [unlimited threads](https://xunit.net/docs/configuration-files#maxParallelThreads). In this mode, the system thread pool will grow (albeit slowly) as needed when it has more work to run than available threads, so such a deadlock should in theory be impossible (though in practice, the system thread pool _probably_ has an actual upper limit, but it's likely so large that it's irrelevant).

**Reply by bforl** [reply]
I removed the static constructor entirely and now have no statics at all.

Same problem. I'm baffled.


"The fact that unrelated tests are frozen would make sense if enough tests are blocking waiting for the lock."


Why would unrelated tests freeze though? They're in different threads and have completely different code.

**Reply by bradwilson** [reply]
You have a limit to the number of tests that can run at the same time (the `maxParallelThreads` setting I linked above). If you have that many tests stuck waiting for a synchronous lock, then literally nothing else can run, because all the threads are blocked because the wait is synchronous rather than asynchronous.

But it's impossible for me to know without seeing the actual situation at hand, without being able to dive into every worker thread and see where they're stuck and why.
